{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from scipy import stats\n",
    "plt.style.use('ggplot')\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "nor_counts=pd.read_table('normalized_counts.txt')\n",
    "col_new=['geneID', 'geneLength', 'geneProduct','185-1','WT-1','185-2','185-3','WT-2','WT-3']\n",
    "nor_counts.columns=col_new\n",
    "nor_counts=nor_counts[['geneID','geneLength','geneProduct','WT-1','WT-2','WT-3','185-1','185-2','185-3']]\n",
    "nor_counts['WT_mean']=np.apply_along_axis(np.mean,1,nor_counts[['WT-1','WT-2','WT-3']])\n",
    "nor_counts['185_mean']=np.apply_along_axis(np.mean,1,nor_counts[['185-1','185-2','185-3']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_id_1=pd.read_table('Gene_id_1_1000.txt')\n",
    "gene_id_2=pd.read_table('Gene_id_1001_2000.txt')\n",
    "gene_id_3=pd.read_table('Gene_id_2001_3000.txt')\n",
    "gene_id_4=pd.read_table('Gene_id_3001_3031.txt')\n",
    "gene_id_main=[gene_id_1,gene_id_2,gene_id_3,gene_id_4]\n",
    "gene_id_main=pd.concat(gene_id_main)\n",
    "gene_id_main=gene_id_main.drop(['Batch1','Genome Name','Unnamed: 7','Genome ID'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "merged=nor_counts.merge(gene_id_main,how='left',left_on='geneID',right_on='Gene ID')\n",
    "df=merged.drop(['Gene Product Name','Gene ID','geneLength'],axis=1)\n",
    "col_order=['geneID','Locus Tag','Gene Symbol','geneProduct','WT-1', 'WT-2', 'WT-3', '185-1', '185-2',\n",
    "       '185-3', 'WT_mean', '185_mean' ]\n",
    "df=df[col_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>Locus Tag</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>geneProduct</th>\n",
       "      <th>WT-1</th>\n",
       "      <th>WT-2</th>\n",
       "      <th>WT-3</th>\n",
       "      <th>185-1</th>\n",
       "      <th>185-2</th>\n",
       "      <th>185-3</th>\n",
       "      <th>WT_mean</th>\n",
       "      <th>185_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>650468844</td>\n",
       "      <td>Clo1313_0001</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>chromosomal replication initiator protein DnaA</td>\n",
       "      <td>95.68</td>\n",
       "      <td>68.16</td>\n",
       "      <td>33.52</td>\n",
       "      <td>12.22</td>\n",
       "      <td>13.38</td>\n",
       "      <td>14.04</td>\n",
       "      <td>65.786667</td>\n",
       "      <td>13.213333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>650468845</td>\n",
       "      <td>Clo1313_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNA polymerase III, beta subunit (EC 2.7.7.7)</td>\n",
       "      <td>91.89</td>\n",
       "      <td>65.97</td>\n",
       "      <td>34.76</td>\n",
       "      <td>7.39</td>\n",
       "      <td>13.24</td>\n",
       "      <td>7.93</td>\n",
       "      <td>64.206667</td>\n",
       "      <td>9.520000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>650468846</td>\n",
       "      <td>Clo1313_0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S4 domain protein YaaA</td>\n",
       "      <td>31.74</td>\n",
       "      <td>23.39</td>\n",
       "      <td>12.33</td>\n",
       "      <td>6.55</td>\n",
       "      <td>7.83</td>\n",
       "      <td>12.05</td>\n",
       "      <td>22.486667</td>\n",
       "      <td>8.810000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>650468847</td>\n",
       "      <td>Clo1313_0004</td>\n",
       "      <td>recF</td>\n",
       "      <td>DNA replication and repair protein RecF</td>\n",
       "      <td>21.31</td>\n",
       "      <td>13.09</td>\n",
       "      <td>9.20</td>\n",
       "      <td>1.22</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.12</td>\n",
       "      <td>14.533333</td>\n",
       "      <td>0.780000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>650468848</td>\n",
       "      <td>Clo1313_0005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>221.06</td>\n",
       "      <td>242.11</td>\n",
       "      <td>267.31</td>\n",
       "      <td>103.35</td>\n",
       "      <td>82.30</td>\n",
       "      <td>130.65</td>\n",
       "      <td>243.493333</td>\n",
       "      <td>105.433333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID     Locus Tag Gene Symbol  \\\n",
       "0  650468844  Clo1313_0001        dnaA   \n",
       "1  650468845  Clo1313_0002         NaN   \n",
       "2  650468846  Clo1313_0003         NaN   \n",
       "3  650468847  Clo1313_0004        recF   \n",
       "4  650468848  Clo1313_0005         NaN   \n",
       "\n",
       "                                      geneProduct    WT-1    WT-2    WT-3  \\\n",
       "0  chromosomal replication initiator protein DnaA   95.68   68.16   33.52   \n",
       "1   DNA polymerase III, beta subunit (EC 2.7.7.7)   91.89   65.97   34.76   \n",
       "2                          S4 domain protein YaaA   31.74   23.39   12.33   \n",
       "3         DNA replication and repair protein RecF   21.31   13.09    9.20   \n",
       "4                            hypothetical protein  221.06  242.11  267.31   \n",
       "\n",
       "    185-1  185-2   185-3     WT_mean    185_mean  \n",
       "0   12.22  13.38   14.04   65.786667   13.213333  \n",
       "1    7.39  13.24    7.93   64.206667    9.520000  \n",
       "2    6.55   7.83   12.05   22.486667    8.810000  \n",
       "3    1.22   0.00    1.12   14.533333    0.780000  \n",
       "4  103.35  82.30  130.65  243.493333  105.433333  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add fold change columns to the data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\miluo\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:1: RuntimeWarning: divide by zero encountered in log2\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n",
      "C:\\Users\\miluo\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:2: RuntimeWarning: divide by zero encountered in log2\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "df['log2(WT)']=np.log2(df['WT_mean'])\n",
    "df['log2(185)']=np.log2(df['185_mean'])\n",
    "df['log2(FC)']=df['log2(185)']-df['log2(WT)']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# remove genes with detectable transcription, which leads to numeric instability in log2 fold change computation.\n",
    "# this filtered data frame should be used in picking out the top 10 gene list\n",
    "# df_filtered=df[np.logical_and(df['WT_mean']>2,df['185_mean']>2)]\n",
    "# df_filtered=df_filtered[['geneID','Locus Tag','Gene Symbol','geneProduct','WT_mean','185_mean','log2(FC)']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows with nan or inf output in log2(FC)\n",
    "row_nan_inf=np.logical_or(np.isnan(df['log2(FC)']),np.isinf(df['log2(FC)']))\n",
    "df=df[~row_nan_inf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>Locus Tag</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>geneProduct</th>\n",
       "      <th>WT-1</th>\n",
       "      <th>WT-2</th>\n",
       "      <th>WT-3</th>\n",
       "      <th>185-1</th>\n",
       "      <th>185-2</th>\n",
       "      <th>185-3</th>\n",
       "      <th>WT_mean</th>\n",
       "      <th>185_mean</th>\n",
       "      <th>log2(WT)</th>\n",
       "      <th>log2(185)</th>\n",
       "      <th>log2(FC)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>650468844</td>\n",
       "      <td>Clo1313_0001</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>chromosomal replication initiator protein DnaA</td>\n",
       "      <td>95.68</td>\n",
       "      <td>68.16</td>\n",
       "      <td>33.52</td>\n",
       "      <td>12.22</td>\n",
       "      <td>13.38</td>\n",
       "      <td>14.04</td>\n",
       "      <td>65.786667</td>\n",
       "      <td>13.213333</td>\n",
       "      <td>6.039723</td>\n",
       "      <td>3.723923</td>\n",
       "      <td>-2.315801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>650468845</td>\n",
       "      <td>Clo1313_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNA polymerase III, beta subunit (EC 2.7.7.7)</td>\n",
       "      <td>91.89</td>\n",
       "      <td>65.97</td>\n",
       "      <td>34.76</td>\n",
       "      <td>7.39</td>\n",
       "      <td>13.24</td>\n",
       "      <td>7.93</td>\n",
       "      <td>64.206667</td>\n",
       "      <td>9.520000</td>\n",
       "      <td>6.004651</td>\n",
       "      <td>3.250962</td>\n",
       "      <td>-2.753690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>650468846</td>\n",
       "      <td>Clo1313_0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S4 domain protein YaaA</td>\n",
       "      <td>31.74</td>\n",
       "      <td>23.39</td>\n",
       "      <td>12.33</td>\n",
       "      <td>6.55</td>\n",
       "      <td>7.83</td>\n",
       "      <td>12.05</td>\n",
       "      <td>22.486667</td>\n",
       "      <td>8.810000</td>\n",
       "      <td>4.490998</td>\n",
       "      <td>3.139142</td>\n",
       "      <td>-1.351856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>650468847</td>\n",
       "      <td>Clo1313_0004</td>\n",
       "      <td>recF</td>\n",
       "      <td>DNA replication and repair protein RecF</td>\n",
       "      <td>21.31</td>\n",
       "      <td>13.09</td>\n",
       "      <td>9.20</td>\n",
       "      <td>1.22</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.12</td>\n",
       "      <td>14.533333</td>\n",
       "      <td>0.780000</td>\n",
       "      <td>3.861294</td>\n",
       "      <td>-0.358454</td>\n",
       "      <td>-4.219748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>650468848</td>\n",
       "      <td>Clo1313_0005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>221.06</td>\n",
       "      <td>242.11</td>\n",
       "      <td>267.31</td>\n",
       "      <td>103.35</td>\n",
       "      <td>82.30</td>\n",
       "      <td>130.65</td>\n",
       "      <td>243.493333</td>\n",
       "      <td>105.433333</td>\n",
       "      <td>7.927738</td>\n",
       "      <td>6.720187</td>\n",
       "      <td>-1.207551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID     Locus Tag Gene Symbol  \\\n",
       "0  650468844  Clo1313_0001        dnaA   \n",
       "1  650468845  Clo1313_0002         NaN   \n",
       "2  650468846  Clo1313_0003         NaN   \n",
       "3  650468847  Clo1313_0004        recF   \n",
       "4  650468848  Clo1313_0005         NaN   \n",
       "\n",
       "                                      geneProduct    WT-1    WT-2    WT-3  \\\n",
       "0  chromosomal replication initiator protein DnaA   95.68   68.16   33.52   \n",
       "1   DNA polymerase III, beta subunit (EC 2.7.7.7)   91.89   65.97   34.76   \n",
       "2                          S4 domain protein YaaA   31.74   23.39   12.33   \n",
       "3         DNA replication and repair protein RecF   21.31   13.09    9.20   \n",
       "4                            hypothetical protein  221.06  242.11  267.31   \n",
       "\n",
       "    185-1  185-2   185-3     WT_mean    185_mean  log2(WT)  log2(185)  \\\n",
       "0   12.22  13.38   14.04   65.786667   13.213333  6.039723   3.723923   \n",
       "1    7.39  13.24    7.93   64.206667    9.520000  6.004651   3.250962   \n",
       "2    6.55   7.83   12.05   22.486667    8.810000  4.490998   3.139142   \n",
       "3    1.22   0.00    1.12   14.533333    0.780000  3.861294  -0.358454   \n",
       "4  103.35  82.30  130.65  243.493333  105.433333  7.927738   6.720187   \n",
       "\n",
       "   log2(FC)  \n",
       "0 -2.315801  \n",
       "1 -2.753690  \n",
       "2 -1.351856  \n",
       "3 -4.219748  \n",
       "4 -1.207551  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compute adjusted p-value (Welch-t-test, Benjamini-Hochberg correction)\n",
    "Using unpooled error to calculate the unadjusted p-value\n",
    "There are several ways to adjust p-values, Bonferroni correction will simply take the unadjusted p-value and times the number of hypothesis. Alternative, FDR-adjusted, aka Benjimini & Hochberg correction, can also be used. For multiple comparison with large m, Bonferroni tend to be too severe. Benjimini & Hochberg method is used herein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calculate_mean_and_variance(df):\n",
    "    \"\"\"\n",
    "    Imput df as a pandas dataframe (m x n).\n",
    "    Each row of df is a different gene of interest, each column of df is a repeat measurement.\n",
    "    Output mean (m x 1) and variance (m x 1).\n",
    "    \"\"\"\n",
    "    df_np=np.array(df)\n",
    "    m=df.shape[0]\n",
    "    n=df.shape[1]\n",
    "    df_mean=np.apply_along_axis(np.mean,1,df_np).reshape((m,1))\n",
    "    df_var=np.sum((df_np-df_mean)**2,axis=1)/n\n",
    "    return df_mean,df_var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calculate_t_stat(interest,control):\n",
    "    \"\"\"\n",
    "    Take in two dataframes of same shape (m x n)\n",
    "    Each row of df is a different gene of interest, each column of df is a repeat measurement.\n",
    "    Calculate the t-stastic (m x 1) and degree of freedom (m x 1), assuming equal sample size, unequal variancies (Welch t-test)\n",
    "    Output in the form of numpy array.\n",
    "    \"\"\"\n",
    "    assert interest.shape==control.shape\n",
    "    m=interest.shape[0]\n",
    "    n=interest.shape[1]\n",
    "    int_mean,int_var=calculate_mean_and_variance(interest)\n",
    "    ctl_mean,ctl_var=calculate_mean_and_variance(control)\n",
    "    variance=np.sqrt((int_var/n)+(ctl_var/n)).reshape((m,1))\n",
    "    t_stat=(int_mean-ctl_mean)/variance\n",
    "    denom=np.power(variance,4)\n",
    "    term1=(int_var/n)**2/(n-1)\n",
    "    term1=term1.reshape((m,1))\n",
    "    term2=(ctl_var/n)**2/(n-1)\n",
    "    term2=term2.reshape((m,1))\n",
    "    deg_of_free=denom/(term1+term2)\n",
    "    return t_stat,deg_of_free\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calculate_p_value(interest,control):\n",
    "    \"\"\"\n",
    "    Take in two dataframes of same shape (m x n), compute unadjusted two-sided p_value (m x 1) using Welch-t-test,\n",
    "    assuming unequal variance.\n",
    "    \"\"\"\n",
    "    t_stat,deg_of_free=calculate_t_stat(interest,control)\n",
    "    p_value=stats.t.sf(np.abs(t_stat),deg_of_free)*2\n",
    "    return p_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calculate_adjusted_p_value(interest,control):\n",
    "    \"\"\"\n",
    "    Take in two dataframes of same shape (m x n), compute unadjusted two-sided p_value (m x 1) using Welch-t-test,\n",
    "    and adjusted p_value (m x 1) using Benjamini-Hochberg method.\n",
    "    \"\"\"\n",
    "    p_value=calculate_p_value(interest,control)\n",
    "    m=len(p_value)\n",
    "    p_tmp=pd.DataFrame(p_value,columns=['p-value'])\n",
    "    p_tmp=p_tmp.sort_values('p-value')\n",
    "    p_rank=stats.rankdata(p_tmp['p-value'])\n",
    "    p_test=p_tmp['p-value']*m/p_rank\n",
    "    p_adj=np.zeros(m)\n",
    "    for i in range(m):\n",
    "        p_adj[i]=np.min([np.min(p_test[i:]),1])   #double min, the outer min ensures p value doesn't exceed 1.\n",
    "    p_tmp['p-adjusted']=p_adj\n",
    "    p_tmp=p_tmp.sort_index()\n",
    "    return np.round(p_tmp['p-value'],3),np.round(p_tmp['p-adjusted'],3)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "results=df[['geneID','Locus Tag','Gene Symbol','geneProduct','WT_mean','185_mean','log2(FC)']]\n",
    "results=results.reset_index(drop=True)\n",
    "results['p-value'],results['p-value adj.']=calculate_adjusted_p_value(df[['185-1','185-2','185-3']],df[['WT-1','WT-2','WT-3']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>Locus Tag</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>geneProduct</th>\n",
       "      <th>WT_mean</th>\n",
       "      <th>185_mean</th>\n",
       "      <th>log2(FC)</th>\n",
       "      <th>p-value</th>\n",
       "      <th>p-value adj.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>650468844</td>\n",
       "      <td>Clo1313_0001</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>chromosomal replication initiator protein DnaA</td>\n",
       "      <td>65.786667</td>\n",
       "      <td>13.213333</td>\n",
       "      <td>-2.315801</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>650468845</td>\n",
       "      <td>Clo1313_0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNA polymerase III, beta subunit (EC 2.7.7.7)</td>\n",
       "      <td>64.206667</td>\n",
       "      <td>9.520000</td>\n",
       "      <td>-2.753690</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>650468846</td>\n",
       "      <td>Clo1313_0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S4 domain protein YaaA</td>\n",
       "      <td>22.486667</td>\n",
       "      <td>8.810000</td>\n",
       "      <td>-1.351856</td>\n",
       "      <td>0.086</td>\n",
       "      <td>0.185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>650468847</td>\n",
       "      <td>Clo1313_0004</td>\n",
       "      <td>recF</td>\n",
       "      <td>DNA replication and repair protein RecF</td>\n",
       "      <td>14.533333</td>\n",
       "      <td>0.780000</td>\n",
       "      <td>-4.219748</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>650468848</td>\n",
       "      <td>Clo1313_0005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>243.493333</td>\n",
       "      <td>105.433333</td>\n",
       "      <td>-1.207551</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.030</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID     Locus Tag Gene Symbol  \\\n",
       "0  650468844  Clo1313_0001        dnaA   \n",
       "1  650468845  Clo1313_0002         NaN   \n",
       "2  650468846  Clo1313_0003         NaN   \n",
       "3  650468847  Clo1313_0004        recF   \n",
       "4  650468848  Clo1313_0005         NaN   \n",
       "\n",
       "                                      geneProduct     WT_mean    185_mean  \\\n",
       "0  chromosomal replication initiator protein DnaA   65.786667   13.213333   \n",
       "1   DNA polymerase III, beta subunit (EC 2.7.7.7)   64.206667    9.520000   \n",
       "2                          S4 domain protein YaaA   22.486667    8.810000   \n",
       "3         DNA replication and repair protein RecF   14.533333    0.780000   \n",
       "4                            hypothetical protein  243.493333  105.433333   \n",
       "\n",
       "   log2(FC)  p-value  p-value adj.  \n",
       "0 -2.315801    0.070         0.166  \n",
       "1 -2.753690    0.054         0.145  \n",
       "2 -1.351856    0.086         0.185  \n",
       "3 -4.219748    0.041         0.127  \n",
       "4 -1.207551    0.001         0.030  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check the fold change in key metabolic pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def check_pathway_transcription(file,df):\n",
    "    pathway=pd.read_table(file)\n",
    "    pathway=pathway[['Gene ID','Locus Tag','Gene Product Name']]\n",
    "    out=pathway.merge(df,how='left',on='Locus Tag')\n",
    "    out=out[['Gene ID','Locus Tag','Gene Product Name','Gene Symbol','WT_mean','185_mean','log2(FC)','p-value','p-value adj.']]\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Locus Tag</th>\n",
       "      <th>Gene Product Name</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>WT_mean</th>\n",
       "      <th>185_mean</th>\n",
       "      <th>log2(FC)</th>\n",
       "      <th>p-value</th>\n",
       "      <th>p-value adj.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>650468944</td>\n",
       "      <td>Clo1313_0099</td>\n",
       "      <td>acetolactate synthase, large subunit (EC 2.2.1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>274.333333</td>\n",
       "      <td>1965.026667</td>\n",
       "      <td>2.840547</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>650468945</td>\n",
       "      <td>Clo1313_0100</td>\n",
       "      <td>acetolactate synthase, small subunit (EC 2.2.1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>295.116667</td>\n",
       "      <td>1796.410000</td>\n",
       "      <td>2.605759</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>650468946</td>\n",
       "      <td>Clo1313_0101</td>\n",
       "      <td>ketol-acid reductoisomerase (EC 1.1.1.86)</td>\n",
       "      <td>ilvC</td>\n",
       "      <td>604.156667</td>\n",
       "      <td>7170.100000</td>\n",
       "      <td>3.568999</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>650469149</td>\n",
       "      <td>Clo1313_0304</td>\n",
       "      <td>dihydroxy-acid dehydratase</td>\n",
       "      <td>ilvD</td>\n",
       "      <td>301.216667</td>\n",
       "      <td>579.973333</td>\n",
       "      <td>0.945185</td>\n",
       "      <td>0.073</td>\n",
       "      <td>0.169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>650469150</td>\n",
       "      <td>Clo1313_0305</td>\n",
       "      <td>acetolactate synthase, large subunit (EC 2.2.1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>370.780000</td>\n",
       "      <td>2555.676667</td>\n",
       "      <td>2.785070</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>650470642</td>\n",
       "      <td>Clo1313_1798</td>\n",
       "      <td>acetaldehyde dehydrogenase (EC 1.2.1.10)/alcoh...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2242.053333</td>\n",
       "      <td>2422.586667</td>\n",
       "      <td>0.111728</td>\n",
       "      <td>0.402</td>\n",
       "      <td>0.520</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene ID     Locus Tag                                  Gene Product Name  \\\n",
       "0  650468944  Clo1313_0099  acetolactate synthase, large subunit (EC 2.2.1...   \n",
       "1  650468945  Clo1313_0100  acetolactate synthase, small subunit (EC 2.2.1...   \n",
       "2  650468946  Clo1313_0101        ketol-acid reductoisomerase (EC 1.1.1.86)     \n",
       "3  650469149  Clo1313_0304                       dihydroxy-acid dehydratase     \n",
       "4  650469150  Clo1313_0305  acetolactate synthase, large subunit (EC 2.2.1...   \n",
       "5  650470642  Clo1313_1798  acetaldehyde dehydrogenase (EC 1.2.1.10)/alcoh...   \n",
       "\n",
       "  Gene Symbol      WT_mean     185_mean  log2(FC)  p-value  p-value adj.  \n",
       "0         NaN   274.333333  1965.026667  2.840547    0.002         0.041  \n",
       "1         NaN   295.116667  1796.410000  2.605759    0.000         0.019  \n",
       "2        ilvC   604.156667  7170.100000  3.568999    0.005         0.054  \n",
       "3        ilvD   301.216667   579.973333  0.945185    0.073         0.169  \n",
       "4         NaN   370.780000  2555.676667  2.785070    0.000         0.017  \n",
       "5         NaN  2242.053333  2422.586667  0.111728    0.402         0.520  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_pathway_transcription('ketoacid_pathway(PWY-7111).txt',results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check_pathway_transcription('mixedacid_pathway(FERMENTATION-PWY).txt',df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check_pathway_transcription(\"reductiveTCAcycle(reductive TCA cycle I).txt\",df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check_pathway_transcription('glycolysis(glycolysis I).txt',df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check_pathway_transcription('carbonhydrate_metabolism.txt',df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check_pathway_transcription('valine_biosynthesis.txt',df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_filtered.sort_values('log2(FC)',ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df[df['Locus Tag']=='Clo1313_0020']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List of most differentially transcribed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>Locus Tag</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>geneProduct</th>\n",
       "      <th>WT_mean</th>\n",
       "      <th>185_mean</th>\n",
       "      <th>log2(FC)</th>\n",
       "      <th>p-value</th>\n",
       "      <th>p-value adj.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>650469348</td>\n",
       "      <td>Clo1313_0501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pectate lyase/Amb allergen</td>\n",
       "      <td>2.946667</td>\n",
       "      <td>89.773333</td>\n",
       "      <td>4.929131</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1976</th>\n",
       "      <td>650471035</td>\n",
       "      <td>Clo1313_2199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>iron-sulfur binding protein</td>\n",
       "      <td>0.573333</td>\n",
       "      <td>15.196667</td>\n",
       "      <td>4.728237</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2138</th>\n",
       "      <td>650471237</td>\n",
       "      <td>Clo1313_2427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transport system permease protein</td>\n",
       "      <td>181.866667</td>\n",
       "      <td>3234.080000</td>\n",
       "      <td>4.152402</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>650469305</td>\n",
       "      <td>Clo1313_0458</td>\n",
       "      <td>rpsZ</td>\n",
       "      <td>ribosomal protein S14</td>\n",
       "      <td>26.940000</td>\n",
       "      <td>338.896667</td>\n",
       "      <td>3.653024</td>\n",
       "      <td>0.076</td>\n",
       "      <td>0.174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>650468946</td>\n",
       "      <td>Clo1313_0101</td>\n",
       "      <td>ilvC</td>\n",
       "      <td>ketol-acid reductoisomerase (EC 1.1.1.86)</td>\n",
       "      <td>604.156667</td>\n",
       "      <td>7170.100000</td>\n",
       "      <td>3.568999</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>650469653</td>\n",
       "      <td>Clo1313_0812</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain of unknown function DUF3298</td>\n",
       "      <td>91.910000</td>\n",
       "      <td>806.266667</td>\n",
       "      <td>3.132963</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>650469595</td>\n",
       "      <td>Clo1313_0745</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncharacterized conserved protein UCP033563</td>\n",
       "      <td>35.236667</td>\n",
       "      <td>255.480000</td>\n",
       "      <td>2.858061</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>650468944</td>\n",
       "      <td>Clo1313_0099</td>\n",
       "      <td>NaN</td>\n",
       "      <td>acetolactate synthase, large subunit (EC 2.2.1.6)</td>\n",
       "      <td>274.333333</td>\n",
       "      <td>1965.026667</td>\n",
       "      <td>2.840547</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1975</th>\n",
       "      <td>650471034</td>\n",
       "      <td>Clo1313_2198</td>\n",
       "      <td>hcp</td>\n",
       "      <td>hybrid cluster protein</td>\n",
       "      <td>2.576667</td>\n",
       "      <td>18.070000</td>\n",
       "      <td>2.810019</td>\n",
       "      <td>0.055</td>\n",
       "      <td>0.145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>650469150</td>\n",
       "      <td>Clo1313_0305</td>\n",
       "      <td>NaN</td>\n",
       "      <td>acetolactate synthase, large subunit (EC 2.2.1.6)</td>\n",
       "      <td>370.780000</td>\n",
       "      <td>2555.676667</td>\n",
       "      <td>2.785070</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         geneID     Locus Tag Gene Symbol  \\\n",
       "451   650469348  Clo1313_0501         NaN   \n",
       "1976  650471035  Clo1313_2199         NaN   \n",
       "2138  650471237  Clo1313_2427         NaN   \n",
       "411   650469305  Clo1313_0458        rpsZ   \n",
       "87    650468946  Clo1313_0101        ilvC   \n",
       "713   650469653  Clo1313_0812         NaN   \n",
       "663   650469595  Clo1313_0745         NaN   \n",
       "85    650468944  Clo1313_0099         NaN   \n",
       "1975  650471034  Clo1313_2198         hcp   \n",
       "261   650469150  Clo1313_0305         NaN   \n",
       "\n",
       "                                            geneProduct     WT_mean  \\\n",
       "451                          Pectate lyase/Amb allergen    2.946667   \n",
       "1976                        iron-sulfur binding protein    0.573333   \n",
       "2138                  transport system permease protein  181.866667   \n",
       "411                               ribosomal protein S14   26.940000   \n",
       "87            ketol-acid reductoisomerase (EC 1.1.1.86)  604.156667   \n",
       "713                  Domain of unknown function DUF3298   91.910000   \n",
       "663         Uncharacterized conserved protein UCP033563   35.236667   \n",
       "85    acetolactate synthase, large subunit (EC 2.2.1.6)  274.333333   \n",
       "1975                             hybrid cluster protein    2.576667   \n",
       "261   acetolactate synthase, large subunit (EC 2.2.1.6)  370.780000   \n",
       "\n",
       "         185_mean  log2(FC)  p-value  p-value adj.  \n",
       "451     89.773333  4.929131    0.010         0.070  \n",
       "1976    15.196667  4.728237    0.047         0.137  \n",
       "2138  3234.080000  4.152402    0.006         0.058  \n",
       "411    338.896667  3.653024    0.076         0.174  \n",
       "87    7170.100000  3.568999    0.005         0.054  \n",
       "713    806.266667  3.132963    0.008         0.064  \n",
       "663    255.480000  2.858061    0.002         0.042  \n",
       "85    1965.026667  2.840547    0.002         0.041  \n",
       "1975    18.070000  2.810019    0.055         0.145  \n",
       "261   2555.676667  2.785070    0.000         0.017  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.sort_values('log2(FC)',ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>Locus Tag</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>geneProduct</th>\n",
       "      <th>WT_mean</th>\n",
       "      <th>185_mean</th>\n",
       "      <th>log2(FC)</th>\n",
       "      <th>p-value</th>\n",
       "      <th>p-value adj.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1269</th>\n",
       "      <td>650470260</td>\n",
       "      <td>Clo1313_1416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>response regulator receiver protein</td>\n",
       "      <td>207.856667</td>\n",
       "      <td>1.010000</td>\n",
       "      <td>-7.685090</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1264</th>\n",
       "      <td>650470255</td>\n",
       "      <td>Clo1313_1411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CheA signal transduction histidine kinase</td>\n",
       "      <td>92.906667</td>\n",
       "      <td>0.636667</td>\n",
       "      <td>-7.189100</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1679</th>\n",
       "      <td>650470706</td>\n",
       "      <td>Clo1313_1866</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>109.580000</td>\n",
       "      <td>0.846667</td>\n",
       "      <td>-7.015975</td>\n",
       "      <td>0.055</td>\n",
       "      <td>0.146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014</th>\n",
       "      <td>650471079</td>\n",
       "      <td>Clo1313_2251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>140.786667</td>\n",
       "      <td>1.170000</td>\n",
       "      <td>-6.910858</td>\n",
       "      <td>0.023</td>\n",
       "      <td>0.094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>879</th>\n",
       "      <td>650469832</td>\n",
       "      <td>Clo1313_0991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>two component transcriptional regulator, LuxR ...</td>\n",
       "      <td>387.713333</td>\n",
       "      <td>3.606667</td>\n",
       "      <td>-6.748180</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2620</th>\n",
       "      <td>650471767</td>\n",
       "      <td>Clo1313_2980</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hemerythrin-like metal-binding protein</td>\n",
       "      <td>611.766667</td>\n",
       "      <td>5.883333</td>\n",
       "      <td>-6.700204</td>\n",
       "      <td>0.051</td>\n",
       "      <td>0.141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266</th>\n",
       "      <td>650470257</td>\n",
       "      <td>Clo1313_1413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MCP methyltransferase, CheR-type</td>\n",
       "      <td>107.100000</td>\n",
       "      <td>1.103333</td>\n",
       "      <td>-6.600946</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1028</th>\n",
       "      <td>650469998</td>\n",
       "      <td>Clo1313_1160</td>\n",
       "      <td>ldh</td>\n",
       "      <td>L-lactate dehydrogenase (EC 1.1.1.27)</td>\n",
       "      <td>300.943333</td>\n",
       "      <td>3.333333</td>\n",
       "      <td>-6.496382</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2022</th>\n",
       "      <td>650471087</td>\n",
       "      <td>Clo1313_2259</td>\n",
       "      <td>NaN</td>\n",
       "      <td>methyl-accepting chemotaxis sensory transducer</td>\n",
       "      <td>23.413333</td>\n",
       "      <td>0.296667</td>\n",
       "      <td>-6.302344</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909</th>\n",
       "      <td>650469864</td>\n",
       "      <td>Clo1313_1023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>70.323333</td>\n",
       "      <td>0.936667</td>\n",
       "      <td>-6.230324</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.071</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         geneID     Locus Tag Gene Symbol  \\\n",
       "1269  650470260  Clo1313_1416         NaN   \n",
       "1264  650470255  Clo1313_1411         NaN   \n",
       "1679  650470706  Clo1313_1866         NaN   \n",
       "2014  650471079  Clo1313_2251         NaN   \n",
       "879   650469832  Clo1313_0991         NaN   \n",
       "2620  650471767  Clo1313_2980         NaN   \n",
       "1266  650470257  Clo1313_1413         NaN   \n",
       "1028  650469998  Clo1313_1160         ldh   \n",
       "2022  650471087  Clo1313_2259         NaN   \n",
       "909   650469864  Clo1313_1023         NaN   \n",
       "\n",
       "                                            geneProduct     WT_mean  185_mean  \\\n",
       "1269                response regulator receiver protein  207.856667  1.010000   \n",
       "1264          CheA signal transduction histidine kinase   92.906667  0.636667   \n",
       "1679                               hypothetical protein  109.580000  0.846667   \n",
       "2014                               hypothetical protein  140.786667  1.170000   \n",
       "879   two component transcriptional regulator, LuxR ...  387.713333  3.606667   \n",
       "2620             hemerythrin-like metal-binding protein  611.766667  5.883333   \n",
       "1266                   MCP methyltransferase, CheR-type  107.100000  1.103333   \n",
       "1028              L-lactate dehydrogenase (EC 1.1.1.27)  300.943333  3.333333   \n",
       "2022     methyl-accepting chemotaxis sensory transducer   23.413333  0.296667   \n",
       "909                                hypothetical protein   70.323333  0.936667   \n",
       "\n",
       "      log2(FC)  p-value  p-value adj.  \n",
       "1269 -7.685090    0.003         0.046  \n",
       "1264 -7.189100    0.006         0.059  \n",
       "1679 -7.015975    0.055         0.146  \n",
       "2014 -6.910858    0.023         0.094  \n",
       "879  -6.748180    0.002         0.039  \n",
       "2620 -6.700204    0.051         0.141  \n",
       "1266 -6.600946    0.003         0.049  \n",
       "1028 -6.496382    0.024         0.096  \n",
       "2022 -6.302344    0.024         0.097  \n",
       "909  -6.230324    0.011         0.071  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.sort_values('log2(FC)').head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Making a heat map (creating a customary color map)\n",
    "After some thought, I think heat map is not a good idea to present this data, as there is only one condition. Without showing repeat samples, it will simply be a heat map of one column (pretty much defeats the purpose of a heat map.)\n",
    "I leave the code here just for my future reference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "test_plot=df.iloc[:10,[2,4,5,6,7,8,9]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_names=list(test_plot.iloc[:,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some exercise for making a heat map in ipython notebook. In this data set, as there is only one condition, \n",
    "# it is better to simply present data as fold change instead of showing the heatmap\n",
    "\n",
    "cdict1 = {'red':   ((0.0, 0.0, 0.0),\n",
    "                   (0.5, 0.0, 0.0),\n",
    "                   (1.0, 1.0, 1.0)),\n",
    "\n",
    "         'green': ((0.0, 1.0, 1.0),\n",
    "                   (0.5, 0.0, 0.0),\n",
    "                   (1.0, 0.0, 0.0)),\n",
    "\n",
    "         'blue':  ((0.0, 0.0, 0.0),\n",
    "                   (0.5, 0.0, 0.0),\n",
    "                   (1.0, 0.0, 0.0))\n",
    "        }\n",
    "green_red=LinearSegmentedColormap('green_red',cdict1)\n",
    "fig=plt.figure(figsize=(10,5))\n",
    "plt.imshow(test_plot.iloc[:,1:],cmap=green_red)\n",
    "plt.colorbar()\n",
    "plt.grid(False)\n",
    "ax=fig.gca()\n",
    "ax.set_yticklabels(gene_names)\n",
    "ax.tick_params(labelsize=20)\n",
    "ax.set_yticks(range(10))\n",
    "ax.set_xticks(range(6))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
